tests/rst_style.py

```diff
@@ -0,0 +1,153 @@
+import sys
+
+import sphinxlint
+from mock import patch
+
+
+CUSTOM_RST_DIRECTIVES = [
+    'card', 'cards',  # cards
+    'example', 'exercise',  # custom_admonitions
+    'spoiler',  # spoilers
+    'tab', 'tabs', 'group-tab', 'code-tab',  # sphinx_tabs
+]
+
+ALLOWED_HEADING_CHARS = ['=', '-', '~', '*', '^']  # In the same order as in the guidelines.
+MAIN_HEADING_CHAR = ALLOWED_HEADING_CHARS[0]
+MAIN_HEADING_RE = re.compile(rf'{MAIN_HEADING_CHAR}+\n[^\n]+\n{MAIN_HEADING_CHAR}+\n')
+HEADING_DELIMITER_RE = re.compile(
+    '^(' + '|'.join(rf'\{char}+' for char in ALLOWED_HEADING_CHARS) + ')\n$'
+)
+FORBIDDEN_HEADING_CHARS = [
+    '#', '"', '\'', '+', '`', '@', '!', ',', '.', '/'
+]  # Exhaustive list at https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections.
+FORBIDDEN_HEADING_DELIMITER_RE = re.compile(
+    '^(' + '|'.join(rf'\{char}+' for char in FORBIDDEN_HEADING_CHARS) + ')\n$'
+)
+ALL_HEADING_CHARS = ALLOWED_HEADING_CHARS + FORBIDDEN_HEADING_CHARS
+
+# TODO override unclear tests
+
+@sphinxlint.checker('.rst')
+def check_heading_delimiters_characters(file, lines, options=None):
+    """ Check that heading delimiters use only allowed characters. """
+    for lno, line in enumerate(lines):
+        if FORBIDDEN_HEADING_DELIMITER_RE.search(line):
+            yield lno + 1, f"illegal use of the character {line[0]} in heading delimiters; use" \
+                           f" any of {', '.join(ALLOWED_HEADING_CHARS)} instead"
+
+@sphinxlint.checker('.rst')
+def check_heading_delimiters_order(file, lines, options=None):
+    """ Check that heading delimiters appear in the correct order. """
+    last_delimiter_char_index = -1  # The index of the heading delimiter char in the ordered list.
+    for lno, line in enumerate(lines):
+        if HEADING_DELIMITER_RE.search(line):  # The line is a heading delimiter.
+            delimiter_char = line[0]
+            delimiter_char_index = ALLOWED_HEADING_CHARS.index(delimiter_char)
+            if delimiter_char_index > last_delimiter_char_index + 1:
+                # There is a leap of more than 1 in the chars used for the heading delimiters.
+                last_delimiter_char = ALLOWED_HEADING_CHARS[last_delimiter_char_index] \
+                    if last_delimiter_char_index != -1 else None
+                yield lno + 1, f"the heading delimiter {delimiter_char} is not allowed after a" \
+                               f" heading with {last_delimiter_char} as delimiter; follow this" \
+                               f" order: {', '.join(ALLOWED_HEADING_CHARS)}"
+            last_delimiter_char_index = delimiter_char_index
+
+@sphinxlint.checker('.rst')
+def check_max_one_main_heading(file, lines, options=None):
+    """ Check that there is at most one main heading (h1) per document. """
+    code = "".join(lines)
+    nb_main_headings = len(MAIN_HEADING_RE.findall(code))
+    if nb_main_headings > 1:
+        yield 1, "the document should have only one main heading"
```

Comment:
Shouldn't we also somehow enforce the presence of a h1 if there are headings ?

Reply:
yup